From 71c877b8a9a3e7739f9c826decf575667e80b094 Mon Sep 17 00:00:00 2001
From: Robert Dabrowski <robertx.dabrowski@intel.com>
Date: Fri, 2 Aug 2019 11:42:47 +0200
Subject: [PATCH] PONRTSYS-2466: Add LAN Ethernet PM counters to ethtool

- implement ethtool api for BP and CTP
- use callbacks in HAL
---
 drivers/net/datapath/dpm/Makefile                |  1 +
 drivers/net/datapath/dpm/datapath.h              | 53 ++++++++++++++++++++++++
 drivers/net/datapath/dpm/datapath_api.c          | 38 +++++++++++++++++
 drivers/net/datapath/dpm/datapath_stats.c        | 28 +++++++++++++
 drivers/net/datapath/dpm/datapath_stats.h        | 30 ++++++++++++++
 drivers/net/datapath/dpm/gswip31/datapath_misc.c | 52 +++++++++++++++++++++++
 drivers/net/datapath/dpm/gswip31/datapath_misc.h | 11 +++++
 drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c | 13 ++++++
 include/net/datapath_api.h                       | 26 ++++++++++++
 include/net/datapath_inst.h                      |  7 ++++
 10 files changed, 259 insertions(+)

diff --git a/drivers/net/datapath/dpm/Makefile b/drivers/net/datapath/dpm/Makefile
index 309120d105dd..d7221ce73879 100644
--- a/drivers/net/datapath/dpm/Makefile
+++ b/drivers/net/datapath/dpm/Makefile
@@ -1,4 +1,5 @@
 obj-$(CONFIG_INTEL_DATAPATH) = datapath_api.o datapath_proc_api.o datapath_proc.o  datapath_misc.o datapath_notifier.o datapath_logical_dev.o datapath_instance.o datapath_soc.o datapath_qos.o datapath_proc_qos.o datapath_tx.o
+obj-$(CONFIG_INTEL_DATAPATH) += datapath_stats.o
 
 ifneq ($(CONFIG_INTEL_DATAPATH_HAL_GSWIP32),)
 obj-$(CONFIG_INTEL_DATAPATH) += gswip32/
diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index 7d77fb95b844..17867b89b02e 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -344,6 +344,59 @@ enum QOS_FLAG {
 	QOS_PORT_CFG_SET, /* set qos port config info */
 };
 
+/* Ethtool strings and counter names (based on GSW_RMON_Port_cnt_t) */
+#define DP_ETHTOOL_STAT_REG_DECLARE	\
+	_DP_ETHTOOL(RxExtendedVlanDiscardPkts, nRxExtendedVlanDiscardPkts) \
+	_DP_ETHTOOL(MtuExceedDiscardPkts, nMtuExceedDiscardPkts) \
+	_DP_ETHTOOL(TxUnderSizeGoodPkts, nTxUnderSizeGoodPkts) \
+	_DP_ETHTOOL(TxOversizeGoodPkts, nTxOversizeGoodPkts) \
+	_DP_ETHTOOL(RxGoodPkts, nRxGoodPkts) \
+	_DP_ETHTOOL(RxUnicastPkts, nRxUnicastPkts) \
+	_DP_ETHTOOL(RxBroadcastPkts, nRxBroadcastPkts) \
+	_DP_ETHTOOL(RxMulticastPkts, nRxMulticastPkts) \
+	_DP_ETHTOOL(RxFCSErrorPkts, nRxFCSErrorPkts) \
+	_DP_ETHTOOL(RxUnderSizeGoodPkts, nRxUnderSizeGoodPkts) \
+	_DP_ETHTOOL(RxOversizeGoodPkts, nRxOversizeGoodPkts) \
+	_DP_ETHTOOL(RxUnderSizeErrorPkts, nRxUnderSizeErrorPkts) \
+	_DP_ETHTOOL(RxGoodPausePkts, nRxGoodPausePkts) \
+	_DP_ETHTOOL(RxOversizeErrorPkts, nRxOversizeErrorPkts) \
+	_DP_ETHTOOL(RxAlignErrorPkts, nRxAlignErrorPkts) \
+	_DP_ETHTOOL(RxFilteredPkts, nRxFilteredPkts) \
+	_DP_ETHTOOL(Rx64BytePkts, nRx64BytePkts) \
+	_DP_ETHTOOL(Rx127BytePkts, nRx127BytePkts) \
+	_DP_ETHTOOL(Rx255BytePkts, nRx255BytePkts) \
+	_DP_ETHTOOL(Rx511BytePkts, nRx511BytePkts) \
+	_DP_ETHTOOL(Rx1023BytePkts, nRx1023BytePkts) \
+	_DP_ETHTOOL(RxMaxBytePkts, nRxMaxBytePkts) \
+	_DP_ETHTOOL(TxGoodPkts, nTxGoodPkts) \
+	_DP_ETHTOOL(TxUnicastPkts, nTxUnicastPkts) \
+	_DP_ETHTOOL(TxBroadcastPkts, nTxBroadcastPkts) \
+	_DP_ETHTOOL(TxMulticastPkts, nTxMulticastPkts) \
+	_DP_ETHTOOL(TxSingleCollCount, nTxSingleCollCount) \
+	_DP_ETHTOOL(TxMultCollCount, nTxMultCollCount) \
+	_DP_ETHTOOL(TxLateCollCount, nTxLateCollCount) \
+	_DP_ETHTOOL(TxExcessCollCount, nTxExcessCollCount) \
+	_DP_ETHTOOL(TxCollCount, nTxCollCount) \
+	_DP_ETHTOOL(TxPauseCount, nTxPauseCount) \
+	_DP_ETHTOOL(Tx64BytePkts, nTx64BytePkts) \
+	_DP_ETHTOOL(Tx127BytePkts, nTx127BytePkts) \
+	_DP_ETHTOOL(Tx255BytePkts, nTx255BytePkts) \
+	_DP_ETHTOOL(Tx511BytePkts, nTx511BytePkts) \
+	_DP_ETHTOOL(Tx1023BytePkts, nTx1023BytePkts) \
+	_DP_ETHTOOL(TxMaxBytePkts, nTxMaxBytePkts) \
+	_DP_ETHTOOL(TxDroppedPkts, nTxDroppedPkts) \
+	_DP_ETHTOOL(TxAcmDroppedPkts, nTxAcmDroppedPkts) \
+	_DP_ETHTOOL(RxDroppedPkts, nRxDroppedPkts) \
+	_DP_ETHTOOL(RxGoodBytes, nRxGoodBytes) \
+	_DP_ETHTOOL(RxBadBytes, nRxBadBytes) \
+	_DP_ETHTOOL(TxGoodBytes, nTxGoodBytes)
+
+enum DP_ETHTOOL_STAT {
+#define _DP_ETHTOOL(x, y)	DP_E_STAT_ ## x,
+DP_ETHTOOL_STAT_REG_DECLARE
+#undef _DP_ETHTOOL
+};
+
 struct dev_mib {
 	atomic_t rx_fn_rxif_pkt; /*! received packet counter */
 	atomic_t rx_fn_txif_pkt; /*! transmitted packet counter */
diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index 60bbddce5b59..e6da4fc58df5 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -2279,6 +2279,44 @@ int bp_pmapper_dev_get(int inst, struct net_device *dev)
 	return -1;
 }
 
+/* ethtool statistics support */
+void dp_net_dev_get_ss_stat_strings(struct net_device *dev, u8 *data)
+{
+	dp_subif_t subif = {0};
+	struct inst_info *pinfo = &dp_port_prop[subif.inst].info;
+
+	if (!pinfo->dp_net_dev_get_ss_stat_strings)
+		return;
+
+	pinfo->dp_net_dev_get_ss_stat_strings(dev, data);
+}
+EXPORT_SYMBOL(dp_net_dev_get_ss_stat_strings);
+
+int dp_net_dev_get_ss_stat_strings_count(struct net_device *dev)
+{
+	dp_subif_t subif = {0};
+	struct inst_info *pinfo = &dp_port_prop[subif.inst].info;
+
+	if (!pinfo->dp_net_dev_get_ss_stat_strings_count)
+		return -EOPNOTSUPP;
+
+	return pinfo->dp_net_dev_get_ss_stat_strings_count(dev);
+}
+EXPORT_SYMBOL(dp_net_dev_get_ss_stat_strings_count);
+
+void dp_net_dev_get_ethtool_stats(struct net_device *dev,
+				  struct ethtool_stats *stats, u64 *data)
+{
+	dp_subif_t subif = {0};
+	struct inst_info *pinfo = &dp_port_prop[subif.inst].info;
+
+	if (!pinfo->dp_net_dev_get_ethtool_stats)
+		return;
+
+	pinfo->dp_net_dev_get_ethtool_stats(dev, stats, data);
+}
+EXPORT_SYMBOL(dp_net_dev_get_ethtool_stats);
+
 #ifdef DP_TEST_EXAMPLE
 void test(void)
 {
diff --git a/drivers/net/datapath/dpm/datapath_stats.c b/drivers/net/datapath/dpm/datapath_stats.c
new file mode 100644
index 000000000000..cea5f4aea1e9
--- /dev/null
+++ b/drivers/net/datapath/dpm/datapath_stats.c
@@ -0,0 +1,28 @@
+/*
+ * Copyright (C) Intel Corporation
+ * Author: Robert Dabrowski <robertx.dabrowski@intel.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation.
+ */
+
+#include "datapath.h"
+
+/* Ethtool statistics strings */
+static const char dp_ethtool_gstrings[][ETH_GSTRING_LEN] = {
+#define _DP_ETHTOOL(x, y)	# x,
+	DP_ETHTOOL_STAT_REG_DECLARE
+#undef _DP_ETHTOOL
+};
+
+/* ethtool statistics support */
+int dp_net_dev_get_ss_stat_strings_count_3X(struct net_device *dev)
+{
+	return ARRAY_SIZE(dp_ethtool_gstrings);
+}
+
+void dp_net_dev_get_ss_stat_strings_3X(struct net_device *dev, u8 *data)
+{
+	memcpy(data, *dp_ethtool_gstrings, sizeof(dp_ethtool_gstrings));
+}
diff --git a/drivers/net/datapath/dpm/datapath_stats.h b/drivers/net/datapath/dpm/datapath_stats.h
new file mode 100644
index 000000000000..d0208b439f31
--- /dev/null
+++ b/drivers/net/datapath/dpm/datapath_stats.h
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) Intel Corporation
+ * Author: Robert Dabrowski <robertx.dabrowski@intel.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation.
+ */
+
+#ifndef DATAPATH_STATS_H
+#define DATAPATH_STATS_H
+
+/*!
+ *@brief dp_net_dev_get_ss_stat_strings_count_3X - function used by callback
+ *dp_net_dev_get_ss_stat_strings_count to retrieve number of counters
+ *@param[in] dev: net device
+ *@return number of counters
+ */
+int dp_net_dev_get_ss_stat_strings_count_3X(struct net_device *dev);
+
+/*!
+ *@brief dp_net_dev_get_ss_stat_strings_3X - function used by gswip callback
+ *dp_net_dev_get_ss_stat_strings to retrieve set of counters' names.
+ *NOTE: The set of counters names is common for gswip31 and gdwip32
+ *@param[in] dev: net device
+ *@param[out] data: buffer pointer where the function will copy counters' names
+ */
+void dp_net_dev_get_ss_stat_strings_3X(struct net_device *dev, u8 *data);
+
+#endif
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index 750aa4aed2e8..a2e3067d4958 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -35,6 +35,7 @@
 #include <net/datapath_api_gswip31.h>
 #include "../datapath.h"
 #include "../datapath_instance.h"
+#include "../datapath_stats.h"
 #include "datapath_proc.h"
 #include "datapath_ppv4.h"
 #include "datapath_misc.h"
@@ -2145,6 +2146,12 @@ int register_dp_cap_gswip31(int flag)
 	cap.info.dp_set_bp_attr = dp_set_bp_attr_31;
 	cap.info.dp_rx = dp_rx_31;
 	cap.info.dp_tx = dp_xmit_31;
+	cap.info.dp_net_dev_get_ss_stat_strings =
+		dp_net_dev_get_ss_stat_strings_3X;
+	cap.info.dp_net_dev_get_ss_stat_strings_count =
+		dp_net_dev_get_ss_stat_strings_count_3X;
+	cap.info.dp_net_dev_get_ethtool_stats =
+		dp_net_dev_get_ethtool_stats_31;
 #if IS_ENABLED(CONFIG_INTEL_DATAPATH_HAL_GSWIP31_MIB)
 	cap.info.dp_get_port_vap_mib = dp_get_port_vap_mib_31;
 	cap.info.dp_clear_netif_mib = dp_clear_netif_mib_31;
@@ -2187,3 +2194,48 @@ int register_dp_cap_gswip31(int flag)
 
 	return 0;
 }
+
+/* ethtool statistics support */
+void dp_net_dev_get_ethtool_stats_31(struct net_device *dev,
+				     struct ethtool_stats *stats, u64 *data)
+{
+	GSW_RMON_Port_cnt_t g_counters = { 0 };
+	struct core_ops *ops = gsw_get_swcore_ops(0);
+	dp_subif_t s_subif;
+	uint32_t flags = 0;
+	u64 *data_dst;
+	int ret = 0;
+
+	/* get subif information from net device */
+	ret = dp_get_netif_subifid(dev, NULL, NULL, NULL, &s_subif,
+				   flags);
+	if (ret != DP_SUCCESS) {
+		netdev_err(dev, "dp_get_netif_subifid returned %d\n", ret);
+		/* dev_put(dev); */
+		return;
+	}
+
+	/* dp_subif_t::flag_bp is set for BP */
+	if (s_subif.flag_bp) {
+		/* BP - Bridge Port */
+		g_counters.ePortType = GSW_BRIDGE_PORT;
+		g_counters.nPortId = s_subif.bport;
+	} else {
+		/* CTP - Connection Termination Point */
+		g_counters.ePortType = GSW_CTP_PORT;
+		g_counters.nPortId = s_subif.subif;
+	}
+
+	/* read hw counters */
+	ret = ops->gsw_rmon_ops.RMON_Port_Get(ops, &g_counters);
+	if (ret != DP_SUCCESS) {
+		netdev_err(dev, "RMON_Port_Get returned %d\n", ret);
+		return;
+	}
+
+	data_dst = data;
+#define _DP_ETHTOOL(x, y)	*data_dst++ = g_counters.y;
+	DP_ETHTOOL_STAT_REG_DECLARE
+#undef _DP_ETHTOOL
+}
+
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.h b/drivers/net/datapath/dpm/gswip31/datapath_misc.h
index fc44fb382aba..459ed41eb984 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.h
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.h
@@ -216,4 +216,15 @@ int set_port_lookup_mode_31(int inst, u8 ep, u32 flags);
 int tc_vlan_set_31(struct core_ops *ops, struct dp_tc_vlan *vlan,
 		   struct dp_tc_vlan_info *info,
 		   int flag);
+
+/*!
+ *@brief dp_net_dev_get_ethtool_stats_31 - function used by callback
+ * dp_net_dev_get_ethtool_stats to retrieve counters' values specyfic for
+ * gswip31
+ *@param[in] dev: net device
+ *@param[out] stats: for dumping NIC-specific statistics
+ *@param[out] data: for dumping counters values sequence
+ */
+void dp_net_dev_get_ethtool_stats_31(struct net_device *dev,
+				     struct ethtool_stats *stats, u64 *data);
 #endif
diff --git a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
index 564a69140576..9631924f2232 100644
--- a/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
+++ b/drivers/net/ethernet/lantiq/ltq_eth_drv_xrx500.c
@@ -414,6 +414,9 @@ static void get_strings(struct net_device *netdev, u32 stringset, u8 *data)
 			data += ETH_GSTRING_LEN;
 		}
 		break;
+	case ETH_SS_STATS:
+		dp_net_dev_get_ss_stat_strings(netdev, data);
+		break;
 	default:
 		break;
 	}
@@ -424,11 +427,19 @@ static int get_stringset_count(struct net_device *netdev, int sset)
 	switch (sset) {
 	case ETH_SS_PRIV_FLAGS:
 		return PRIV_FLAGS_STR_LEN;
+	case ETH_SS_STATS:
+		return dp_net_dev_get_ss_stat_strings_count(netdev);
 	default:
 		return -EOPNOTSUPP;
 	}
 }
 
+static void get_ethtool_stats(struct net_device *dev,
+			      struct ethtool_stats *stats, u64 *data)
+{
+	dp_net_dev_get_ethtool_stats(dev, stats, data);
+}
+
 /* Structure of the ether tool operation in Phy case  */
 static const struct ethtool_ops ethtool_ops = {
 	.get_drvinfo		= get_drvinfo,
@@ -444,6 +455,7 @@ static const struct ethtool_ops ethtool_ops = {
 	.get_sset_count		= get_stringset_count,
 	.get_priv_flags		= get_priv_flags,
 	.set_priv_flags		= set_priv_flags,
+	.get_ethtool_stats	= get_ethtool_stats,
 };
 
 /* Structure of the ether tool operation in No-Phy case */
@@ -455,6 +467,7 @@ static const struct ethtool_ops serdes_ethtool_ops = {
 	.get_sset_count		= get_stringset_count,
 	.get_priv_flags		= get_priv_flags,
 	.set_priv_flags		= set_priv_flags,
+	.get_ethtool_stats	= get_ethtool_stats,
 };
 
 /* open the network device interface*/
diff --git a/include/net/datapath_api.h b/include/net/datapath_api.h
index 93f6fed5bb67..3e4af28960ef 100644
--- a/include/net/datapath_api.h
+++ b/include/net/datapath_api.h
@@ -1879,4 +1879,30 @@ struct dp_bp_attr {
  */
 int dp_set_bp_attr(struct dp_bp_attr *conf, uint32_t flag);
 
+/*!
+ *@brief dp_net_dev_get_ss_stat_strings API - function used by ethtool
+ *	 to retrieve set of counters' names
+ *@param[in] dev: net device
+ *@param[out] data: buffer pointer where the function will copy counters' names
+ */
+void dp_net_dev_get_ss_stat_strings(struct net_device *dev, u8 *data);
+
+/*!
+ *@brief dp_net_dev_get_ss_stat_strings_count API - function used by ethtool
+ *	 to retrieve number of counters
+ *@param[in] dev: net device
+ *@return number of counters
+ */
+int dp_net_dev_get_ss_stat_strings_count(struct net_device *dev);
+
+/*!
+ *@brief dp_net_dev_get_ethtool_stats API - function used by ethtool
+ *	 to retrieve counters' values
+ *@param[in] dev: net device
+ *@param[out] stats: for dumping NIC-specific statistics
+ *@param[out] data: for dumping counters values sequence
+ */
+void dp_net_dev_get_ethtool_stats(struct net_device *dev,
+				  struct ethtool_stats *stats, u64 *data);
+
 #endif /*DATAPATH_API_H */
diff --git a/include/net/datapath_inst.h b/include/net/datapath_inst.h
index 5b45b1311e14..f567a53caad2 100644
--- a/include/net/datapath_inst.h
+++ b/include/net/datapath_inst.h
@@ -134,6 +134,13 @@ struct inst_info {
 	int32_t (*dp_tx)(struct net_device *rx_if, dp_subif_t *rx_subif,
 			 struct sk_buff *skb, int32_t len, uint32_t flags);
 
+	void (*dp_net_dev_get_ss_stat_strings)(struct net_device *dev,
+					       u8 *data);
+	int (*dp_net_dev_get_ss_stat_strings_count)(struct net_device *dev);
+	void (*dp_net_dev_get_ethtool_stats)(struct net_device *dev,
+					     struct ethtool_stats *stats,
+					     u64 *data);
+
 #if IS_ENABLED(CONFIG_INTEL_DATAPATH_SWITCHDEV)
 	int swdev_flag;
 	int (*swdev_alloc_bridge_id)(int inst);
